builtin/consul/config_sourcer.go

```diff
@@ -0,0 +1,410 @@
+			reqLogger = reqLogger.With("datacenter", kvReq.Partition)
+		}
+
+		cacheVal, ok := cs.cache[kvReq.cacheKey()]
+		if !ok {
+			reqLogger.Trace("querying Consul KV")
+			kvpair, meta, err := cs.client.Get(kvReq.Key, opts.WithContext(ctx))
+			if err != nil {
+				result.Result = &pb.ConfigSource_Value_Error{
+					Error: status.New(codes.Aborted, err.Error()).Proto(),
+				}
+
+				continue
+			}
+
+			refreshCtx, cancel := context.WithCancel(context.Background())
+			cacheVal = &cachedVal{
+				kvPair: kvpair,
+				cancel: cancel,
+				err:    nil,
+			}
+			cs.cache[kvReq.cacheKey()] = cacheVal
+			cs.startBlockingQuery(refreshCtx, log, kvReq, meta.LastIndex, opts)
+		}
+
+		if cacheVal.err != nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.Aborted, cacheVal.err.Error()).Proto(),
+			}
+		}
+
+		if cacheVal.kvPair == nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.NotFound, fmt.Sprintf("Configuration for key %s doesn't exist", kvReq.Key)).Proto(),
+			}
+
+			continue
+		}
+
+		result.Result = &pb.ConfigSource_Value_Value{
+			Value: string(cacheVal.kvPair.Value),
+		}
+	}
+
+	return results, nil
+}
+
+func (cs *ConfigSourcer) startBlockingQuery(ctx context.Context, logger hclog.Logger, kvReq reqConfig, lastIndex uint64, opts *api.QueryOptions) {
+	go func() {
+		opts := opts.WithContext(ctx)
+		failures := 0
+		// Ideally we would use the github.com/hashicorp/consul/api/watch package. However that package doesn't support
+		// namespaces and partitions except with the global client defaults and thus isn't suitable for this usage.
+
+		for {
+			logger.Trace("Issuing blocking query", "wait-index", lastIndex)
+			// set the wait index to use for the query
+			opts.WaitIndex = lastIndex
+			pair, meta, err := cs.client.Get(kvReq.Key, opts)
+
+			// check if we are being stopped
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// KV entry not updated - do nothing
+			if meta != nil && meta.LastIndex == lastIndex {
+				logger.Trace("KV value unchanged")
+				continue
+			}
+
+			// update the data within our cache
+			cs.mu.Lock()
+			val, ok := cs.cache[kvReq.cacheKey()]
+			if !ok {
+				logger.Error("KV data is not present within the cache - stopping blocking query refresher")
+				cs.mu.Unlock()
+				return
+			}
+			val.kvPair = pair
+			val.err = err
+			cs.mu.Unlock()
+
+			// now determine the next wait index
+			if err == nil {
+				lastIndex = meta.LastIndex
+				failures = 0
+			} else {
+				// reset the last index to 0 to do a non-blocking query
+				lastIndex = 0
+				// Set up for exponental backoff
+				failures++
+
+				retry := retryInterval * time.Duration(failures*failures)
+				if retry > maxBackoffTime {
+					retry = maxBackoffTime
+				}
+				logger.Error("KV Get errored", "error", err, "retry", retry.String())
+				select {
+				case <-time.After(retry):
+				case <-ctx.Done():
+					return
+				}
+			}
```

Comment:
Nit: reduce `if err == nil {...} else {...}` to just `if err != nil {...}`, less mental stack to keep track of when reading the code. Just a suggestion 

```suggestion
			if err != nil {
				// reset the last index to 0 to do a non-blocking query
				lastIndex = 0
				// Set up for exponental backoff
				failures++

				retry := retryInterval * time.Duration(failures*failures)
				if retry > maxBackoffTime {
					retry = maxBackoffTime
				}
				logger.Error("KV Get errored", "error", err, "retry", retry.String())
				select {
				case <-time.After(retry):
				        continue
				case <-ctx.Done():
					return
				}
			}
			lastIndex = meta.LastIndex
			failures = 0
```